File: src/mVasu.Pwa/src/app/core/models/tiskilista-distinct-values.dto.ts
/**
 * Mirrors mVasu.Api.Contracts.TiskilistaDistinctValuesDto. Returned by
 * GET /api/tiskilista/distinct-values to populate filter dropdowns —
 * scoped to the user's BranchCode visibility, no other filters applied.
 */
export interface TiskilistaKuntaKaupunginosaDto {
  readonly kunta: string;
  readonly kaupunginosa: string;
}

export interface TiskilistaDistinctValuesDto {
  readonly lajit: ReadonlyArray<string>;
  readonly tyypit: ReadonlyArray<string>;
  readonly kunnat: ReadonlyArray<string>;
  readonly kaupunginosat: ReadonlyArray<string>;
  readonly kaupunginosatByKunta: ReadonlyArray<TiskilistaKuntaKaupunginosaDto>;
  readonly sopimustilat: ReadonlyArray<string>;
  readonly isannoitsijat: ReadonlyArray<string>;
  readonly markkinoijat: ReadonlyArray<string>;
  readonly tilat: ReadonlyArray<string>;
}

export const EMPTY_TISKILISTA_DISTINCT_VALUES: TiskilistaDistinctValuesDto = {
  lajit: [],
  tyypit: [],
  kunnat: [],
  kaupunginosat: [],
  kaupunginosatByKunta: [],
  sopimustilat: [],
  isannoitsijat: [],
  markkinoijat: [],
  tilat: [],
};

File: src/mVasu.Api.Contracts/TiskilistaDistinctValuesDto.cs
namespace mVasu.Api.Contracts;

/// <summary>
/// Distinct values across the user's visible Tiskilista rows for the
/// dimensions the PWA exposes as multi-select filter dropdowns.
/// Scope: <c>UserAreaScope</c> (BranchCode), no other filters applied —
/// the dropdown should offer every value the user can ever see.
/// </summary>
public sealed record TiskilistaDistinctValuesDto(
    IReadOnlyList<string> Lajit,
    IReadOnlyList<string> Tyypit,
    IReadOnlyList<string> Kunnat,
    IReadOnlyList<string> Kaupunginosat,
    IReadOnlyList<TiskilistaKuntaKaupunginosaDto> KaupunginosatByKunta,
    IReadOnlyList<string> Sopimustilat,
    IReadOnlyList<string> Isannoitsijat,
    IReadOnlyList<string> Markkinoijat,
    IReadOnlyList<string> Tilat);

/// <summary>
/// Distinct (kunta, kaupunginosa) pair so the PWA can scope the
/// Kaupunginosa multi-select dropdown to the kunta the user already
/// selected. Sent in addition to the flat Kaupunginosat list so the
/// "no kunta selected" case still has every district available.
/// </summary>
public sealed record TiskilistaKuntaKaupunginosaDto(
    string Kunta,
    string Kaupunginosa);
